crates/gossip/src/storage_arc.rs

```diff
@@ -0,0 +1,290 @@
+use kitsune2_api::agent::{DynLocalAgent, LocalAgent};
+use kitsune2_api::{DhtArc, K2Result};
+use kitsune2_dht::snapshot::DhtSnapshot;
+use kitsune2_dht::ArcSet;
+use std::collections::HashSet;
+
+/// Update the storage arcs of local agents based on a DhtSnapshot.
+///
+/// The snapshot may either be a [DhtSnapshot::RingSectorDetails] or a [DhtSnapshot::Minimal].
+/// A minimal snapshot may only be provided when the snapshot has already been checked to be
+/// identical to our own minimal snapshot. In that case we know that the entire common arc set
+/// was in sync.
+///
+/// Note that the DHT model prioritises syncing the disc. So if we've reached a ring diff then we
+/// can assume that the disc was synced at the point this diff was produced. This function requires
+/// the input [DhtSnapshot] to be a [DhtSnapshot::RingSectorDetails] variant.
+///
+/// The function looks for sectors within the common arc set that did not show up as a mismatch
+/// in the snapshot. Such sectors are considered synced. If any synced sectors can be added to the
+/// start or end of the current storage arc for a local agent, it will do so. Any synced sectors
+/// that cannot be added to the storage arc to produce a larger continuous arc are ignored.
+pub(crate) fn update_storage_arcs(
```

Comment:
> If any synced sectors can be added to the ***start or*** end of the current storage arc for a local agent, it will do so.

In legacy, storage arcs must start in the block that contains the agent's loc. Are we considering changing that here?